cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/DefaultSparkExecutionContext.scala

```diff
@@ -141,41 +143,57 @@ class DefaultSparkExecutionContext(runtimeContext: SparkRuntimeContext,
                                       (implicit decoder: StreamEvent => T): RDD[T] = {
-    val streamId: Id.Stream = Id.Stream.from(getNamespace, streamName)
-
-    // Clone the configuration since it's dataset specification and shouldn't affect the global hConf
-    val configuration = configureStreamInput(new Configuration(runtimeContext.getConfiguration),
-                                             streamId, startTime, endTime)
-    val rdd = sc.newAPIHadoopRDD(configuration, classOf[StreamInputFormat[LongWritable, StreamEvent]],
-                                 classOf[LongWritable], classOf[StreamEvent])
-
-    // Register for stream usage for the Spark program
-    val oldProgramId = runtimeContext.getProgram.getId
-    val owners = List(oldProgramId)
-    try {
-      runtimeContext.getStreamAdmin.register(owners, streamId)
-      runtimeContext.getStreamAdmin.addAccess(new Id.Run(oldProgramId, getRunId.getId), streamId, AccessType.READ)
-    }
-    catch {
-      case e: Exception =>
-        LOG.warn("Failed to registry usage of {} -> {}", streamId, owners, e)
-    }
+    val rdd: RDD[(Long, StreamEvent)] = fromStream(sc, streamName, startTime, endTime, None)
 
     // Wrap the StreamEvent with a SerializableStreamEvent
     // Don't use rdd.values() as it brings in implicit object from SparkContext, which is not available in Spark 1.2
     rdd.map(t => new SerializableStreamEvent(t._2)).map(decoder)
   }
 
+  override def fromStream[T: ClassTag](sc: SparkContext, streamName: String, formatSpec: FormatSpecification,
+                                       startTime: Long, endTime: Long): RDD[(Long, GenericStreamEventData[T])] = {
+    fromStream(sc, streamName, startTime, endTime, Some(formatSpec))
+  }
+
+  /**
+    * Creates a [[org.apache.spark.rdd.RDD]] by reading from the given stream and time range.
+    *
+    * @param sc the [[org.apache.spark.SparkContext]] to use
+    * @param streamName name of the stream
+    * @param startTime  the starting time of the stream to be read in milliseconds (inclusive);
+    *                   passing in `0` means start reading from the first event available in the stream.
+    * @param endTime the ending time of the streams to be read in milliseconds (exclusive);
+    *                passing in `Long#MAX_VALUE` means read up to latest event available in the stream.
+    * @param formatSpec if provided, it describes the format in the stream and will be used to decode stream events
+    *                   to the given value type `T`
+    * @return a new [[org.apache.spark.rdd.RDD]] instance that reads from the given stream.
+    */
+  private def fromStream[T: ClassTag](sc: SparkContext, streamName: String,
+                                      startTime: Long, endTime: Long,
+                                      formatSpec: Option[FormatSpecification]): RDD[(Long, T)] = {
+    val streamId = new StreamId(getNamespace, streamName)
```

Comment:
Why need to have all those factory methods for single argument constructor? Factory methods are useful if you have many different arguments to the constructor. Also separating the factory method from the class itself is not very good because it makes one has no clue on where to find them (how do I know to use `Ids` to construct an instance of `NamespaceId`)?